_posts/2025/2025-02-19-using-deep_dup-to-avoid-surprises.md

```diff
@@ -0,0 +1,96 @@
+---
+title: "Using Deep Duplication to Avoid Surprises"
+description: "Using deep_dup for comprehensive object copying"
+layout: article
+category: ruby
+date: 2025-02-19 01:00
+image:
+  base: "2024/using-deep_dup-to-avoid-surprises"
+  alt: "Repeated windows in a office building"
+  credit: "Rex Krithiran"
+  source: "https://unsplash.com/photos/a-black-and-white-image-of-a-mans-face-DkU8j3N4xak"
+---
+
+When working with complex data structures in Rails, you might need to create a copy of an object that includes all its nested objects. But beware if you’re copying a complex structure.
+
+Rails provides `deep_dup`, offering a powerful way to create deep copies of object structures.
+
+## Instead of…
+
+...using regular duplication methods which do not copy nested structures
+
+```ruby
+og_andy = {
+  name: "Andy",
+  age: 45,
+  address: { city: "Brighton", country: "UK" },
+  hobbies: ["football", "parenting"]
+}
+
+new_andy = og_andy.dup
+
+# see [thankgoodness.game](https://thankgoodness.game)
+new_andy[:address][:city] = "Barnsworth"
+new_andy[:hobbies] << "slapping"
+new_andy[:hobbies] << "jumping"
+
+og_andy[:address][:city]
+#=> "Barnsworth"
+og_andy[:hobbies]
+#=> ["football", "parenting", "slapping", "jumping"]
+```
+
+## Use…
+
+...Rails's `deep_dup` method for comprehensive copying:
+
+```ruby
+og_andy = {
+  name: "Andy",
+  age: 45,
+  address: { city: "Brighton", country: "UK" },
+  hobbies: ["football", "parenting"]
+}
+
+new_andy = og_andy.deep_dup
+
+new_andy[:address][:city] = "Barnsworth"
+new_andy[:hobbies] << "slapping"
+new_andy[:hobbies] << "jumping"
+
+og_andy[:address][:city]
+#=> "Brighton"
+og_andy[:hobbies]
+#=> ["football", "parenting"]
+new_andy[:address][:city]
+#=> "Barnsworth"
+new_andy[:hobbies]
+#=> ["football", "parenting", "slapping", "jumping"]
+```
+
+## Why?
+
+The `deep_dup` method, provided by Active Support, creates a fully independent copy of the original object, with all nested structures also duplicated; including nested hashes, arrays, and other objects.
+
+This fully-duplicated structure prevents unintended modifications, as operating on the copy means changes don't affect the original, avoiding subtle bugs—like the above—from occurring in your code.
+
+Note that Rails’s implementation correctly handles custom objects that implement their own `#dup` method.
```

Comment:
Unclear what this sentence is trying to say.

Reply:
```suggestion
Most objects implement the `#dup` method as it's part of [Ruby's core method set on `Object`](https://docs.ruby-lang.org/en/master/Object.html#method-i-dup). Rails’s implementation correctly utilises any Ruby objects that implement their own `#dup` method.
```